### ISSUE TITLE
Tabs widget exhibits odd behavior when it's on a page that is linked to

### ISSUE THREAD

**Issue by bkosborne** [jquery/jquery-mobile#7169]
A bit of a confusing title, sorry.

When a tabs widget is on a page, and I link to that page, it seems that the page is loaded twice. The second time it's loaded, almost the entire page is re-inserted into the DOM near the original tabs widget for the first page load.

Still confused?

[Here's a working demonstration](http://bkosborne.com/jquery_mobile_tabs/page1.html). Notice how the page loads twice when clicking the link. If you go to `page2.html` directly, there is no issue. Only when the page is loaded via AJAX thru JQM.

The example code is taken almost exactly from the [docs page](http://demos.jquerymobile.com/1.4.1/tabs/).

I thought maybe the `navbar` widget was causing some issues, so I removed that, but it makes no difference.

I have a [stack overflow post](http://stackoverflow.com/questions/21864959) as well as a [jquery mobile forums post](https://forum.jquery.com/topic/when-linking-to-a-page-with-a-tabs-widget-the-page-is-loaded-twice) on the subject.

I'm either missing something very obvious, or I believe there is a bug in the implementation of the tabs widget or the documentation for the tabs widget.

**Comment by slavap** [comment]
Workaround proposed by @gabrielschulhof is almost OK for me, but I've encounter some issues in IE9, for some reason pagebeforeshow was not raised in some cases. So I've improved the code with setInterval(), and now it works just fine in all browsers:
```javascript
$.widget( "ui.tabs", $.ui.tabs, {

    _createWidget: function( options, element ) {
        var page, delayedCreate,
            that = this;

        if ( $.mobile.page ) {
            page = $( element )
                .parents( ":jqmData(role='page'),:mobile-page" )
                .first();

            if ( page.length > 0 && !page.hasClass( "ui-page-active" ) ) {
                delayedCreate = this._super;
                page.one( "pagebeforeshow", function() {
                    //if (window.console) console.log("pagebeforeshow: delayedCreate exec");
                    if (delayedCreate !== null) delayedCreate.call( that, options, element );
                    delayedCreate = null;
                });
                var wait = setInterval(function() {
                    if (delayedCreate === null) clearInterval(wait);
                    else {
                        if ( page.hasClass( "ui-page-active" ) ) {
                            if (delayedCreate !== null) delayedCreate.call( that, options, element );
                            delayedCreate = null;
                            clearInterval(wait);
                        }
                    }
                }, 200);
            } else {
                return this._super( options, element );
            }
        } else {
            return this._super( options, element );
        }
    }
}); 
```

**Comment by slavap** [comment]
I've backported the latest @gabrielschulhof fix to jqm 1.4.5, so now tabs is working just fine.
_processTabs() and load() are just copies of original 1.4.5 code with isLocal() replaced by that._isLocal()

```javascript
$.widget( "ui.tabs", $.ui.tabs, {

    _isLocal: function( anchor ) {
        var path, baseUrl, absUrl;

        if ( $.mobile.ajaxEnabled ) {
            path = $.mobile.path;
            baseUrl = path.parseUrl( $.mobile.base.element.attr( "href" ) );
            absUrl = path.parseUrl( path.makeUrlAbsolute( anchor.getAttribute( "href" ),
                baseUrl ) );

            return ( path.isSameDomain( absUrl.href, baseUrl.href ) &&
                absUrl.pathname === baseUrl.pathname );
        }

        var rhash = /#.*$/;
        var anchorUrl, locationUrl;

        anchorUrl = anchor.href.replace( rhash, "" );
        locationUrl = location.href.replace( rhash, "" );

        // decoding may throw an error if the URL isn't UTF-8 (#9518)
        try {
            anchorUrl = decodeURIComponent( anchorUrl );
        } catch ( error ) {}
        try {
            locationUrl = decodeURIComponent( locationUrl );
        } catch ( error ) {}

        return anchor.hash.length > 1 && anchorUrl === locationUrl;
    },

    _processTabs: function() {
        var that = this;

        this.tablist = this._getList()
            .addClass( "ui-tabs-nav ui-helper-reset ui-helper-clearfix ui-widget-header ui-corner-all" )
            .attr( "role", "tablist" );

        this.tabs = this.tablist.find( "> li:has(a[href])" )
            .addClass( "ui-state-default ui-corner-top" )
            .attr({
                role: "tab",
                tabIndex: -1
            });

        this.anchors = this.tabs.map(function() {
                return $( "a", this )[ 0 ];
            })
            .addClass( "ui-tabs-anchor" )
            .attr({
                role: "presentation",
                tabIndex: -1
            });

        this.panels = $();

        this.anchors.each(function( i, anchor ) {
            var selector, panel, panelId,
                anchorId = $( anchor ).uniqueId().attr( "id" ),
                tab = $( anchor ).closest( "li" ),
                originalAriaControls = tab.attr( "aria-controls" );

            // inline tab
            if ( that._isLocal( anchor ) ) {
                selector = anchor.hash;
                panel = that.element.find( that._sanitizeSelector( selector ) );
            // remote tab
            } else {
                panelId = that._tabId( tab );
                selector = "#" + panelId;
                panel = that.element.find( selector );
                if ( !panel.length ) {
                    panel = that._createPanel( panelId );
                    panel.insertAfter( that.panels[ i - 1 ] || that.tablist );
                }
                panel.attr( "aria-live", "polite" );
            }

            if ( panel.length) {
                that.panels = that.panels.add( panel );
            }
            if ( originalAriaControls ) {
                tab.data( "ui-tabs-aria-controls", originalAriaControls );
            }
            tab.attr({
                "aria-controls": selector.substring( 1 ),
                "aria-labelledby": anchorId
            });
            panel.attr( "aria-labelledby", anchorId );
        });

        this.panels
            .addClass( "ui-tabs-panel ui-widget-content ui-corner-bottom" )
            .attr( "role", "tabpanel" );
    },

    load: function( index, event ) {
        index = this._getIndex( index );
        var that = this,
            tab = this.tabs.eq( index ),
            anchor = tab.find( ".ui-tabs-anchor" ),
            panel = this._getPanelForTab( tab ),
            eventData = {
                tab: tab,
                panel: panel
            };

        // not remote
        if ( that._isLocal( anchor[ 0 ] ) ) {
            return;
        }

        this.xhr = $.ajax( this._ajaxSettings( anchor, event, eventData ) );

        // support: jQuery <1.8
        // jQuery <1.8 returns false if the request is canceled in beforeSend,
        // but as of 1.8, $.ajax() always returns a jqXHR object.
        if ( this.xhr && this.xhr.statusText !== "canceled" ) {
            tab.addClass( "ui-tabs-loading" );
            panel.attr( "aria-busy", "true" );

            this.xhr
                .success(function( response ) {
                    // support: jQuery <1.8
                    // http://bugs.jquery.com/ticket/11778
                    setTimeout(function() {
                        panel.html( response );
                        that._trigger( "load", event, eventData );
                    }, 1 );
                })
                .complete(function( jqXHR, status ) {
                    // support: jQuery <1.8
                    // http://bugs.jquery.com/ticket/11778
                    setTimeout(function() {
                        if ( status === "abort" ) {
                            that.panels.stop( false, true );
                        }

                        tab.removeClass( "ui-tabs-loading" );
                        panel.removeAttr( "aria-busy" );

                        if ( jqXHR === that.xhr ) {
                            delete that.xhr;
                        }
                    }, 1 );
                });
        }
    }

}); 
```

**Comment by arschmitz** [comment]
This is already fixed on 1.5-dev

**Comment by gizmore** [comment]
Here is a fix for 1.5.0 if local tabs trigger an ajax load even though $.mobile.ajaxEnabled is set to false.

    /** Fix for local tabs **/
    $.widget( "ui.tabs", $.ui.tabs, {
    
    	_isLocal: function( anchor ) {
    		var path, baseUrl, absUrl;
    
    //		if ( $.mobile.ajaxEnabled ) { // Here is the bug :)
    			path = $.mobile.path;
    			baseUrl = path.parseUrl( $.mobile.base.element().attr( "href" ) );
    			absUrl = path.parseUrl( path.makeUrlAbsolute( anchor.getAttribute( "href" ),
    				baseUrl ) );
    
    			return ( path.isSameDomain( absUrl.href, baseUrl.href ) &&
    				absUrl.pathname === baseUrl.pathname );
    //		}
    
    		return this._superApply( arguments );
    	}
    
    } );

Put this directly after including jqm
like
    <script src="jquerymobile.js">
    <script>
    thefix
    </script>
else the patch is applied too late.

Enjoy in 2021!
